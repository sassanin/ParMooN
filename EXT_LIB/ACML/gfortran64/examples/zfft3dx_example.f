*     ZFFT3DX Example Program Text
*
*     ACML version 2.5 Copyright AMD,NAG 2004
*
C     .. Parameters ..
      INTEGER          LMAX, MMAX, NMAX, LW
      PARAMETER        (LMAX=10,MMAX=10,NMAX=10,
     *                 LW=9*MMAX+LMAX*MMAX*NMAX)
C     .. Local Scalars ..
      DOUBLE PRECISION SCALE
      INTEGER          I, INFO, IX, IY, J, K, L, M, N
      LOGICAL          INPL, LTRANS
C     .. Local Arrays ..
      COMPLEX *16      COMM(LW), X(LMAX*MMAX*NMAX), XX(LMAX*MMAX*NMAX),
     *                 Y(LMAX*MMAX*NMAX)
C     .. External Subroutines ..
      EXTERNAL         ZFFT3DX
C     .. Intrinsic Functions ..
      INTRINSIC        COS, DBLE, DCMPLX, DSIN
C     .. Executable Statements ..
*
      WRITE (*,FMT=99998)
     *  'ACML example: FFT of a complex sequence using ZFFT3DX'
      WRITE (*,FMT=99998)
     *  '-----------------------------------------------------'
      WRITE (*,FMT=*)
*
*     The sequence of complex data
      L = 5
      M = 2
      N = 2
      SCALE = 1.0D0
      INPL = .FALSE.
      LTRANS = .FALSE.
      IX = 1
      DO 60 I = 1, N
         DO 40 J = 1, M
            DO 20 K = 1, L
               X(IX) = DCMPLX(1.0D0+COS(DBLE(IX)),DSIN(DBLE(IX)))
               XX(IX) = X(IX)
               IX = IX + 1
   20       CONTINUE
   40    CONTINUE
   60 CONTINUE
*
*     Initialize communication work array
      CALL ZFFT3DX(0,SCALE,LTRANS,INPL,L,M,N,X,Y,COMM,INFO)
*
*     Compute a forward transform
      CALL ZFFT3DX(-1,SCALE,LTRANS,INPL,L,M,N,X,Y,COMM,INFO)
*
      SCALE = 1.0D0/DBLE(L*M*N)
      WRITE (*,FMT=99998) 'Components of discrete Fourier transform:'
      WRITE (*,FMT=*)
      IF (LTRANS) THEN
         IY = 1
         DO 120 I = 1, N
            WRITE (*,FMT=*)
            WRITE (*,FMT=99997) 'Third dimension index = ', I
            DO 100 J = 1, M
               WRITE (*,FMT=99997) ' Column Number ', J
               WRITE (*,FMT=99998) '          Real    Imag'
               DO 80 K = 1, L
                  WRITE (*,FMT=99999) K, Y(IY)
                  IY = IY + 1
   80          CONTINUE
  100       CONTINUE
  120    CONTINUE
         CALL ZFFT3DX(1,SCALE,LTRANS,INPL,L,M,N,Y,X,COMM,INFO)
      ELSE
         DO 180 I = 1, N
            WRITE (*,FMT=*)
            WRITE (*,FMT=99997) 'Third dimension index = ', I
            IX = I
            DO 160 J = 1, M
               WRITE (*,FMT=99997) ' Column Number ', J
               WRITE (*,FMT=99998) '          Real    Imag'
               IY = IX
               DO 140 K = 1, L
                  WRITE (*,FMT=99999) K, Y(IY)
                  IY = IY + N*M
  140          CONTINUE
               IX = IX + N
  160       CONTINUE
  180    CONTINUE
         CALL ZFFT3DX(1,SCALE,LTRANS,INPL,N,M,L,Y,X,COMM,INFO)
      END IF
*
*     Compute the reverse transform
*
      WRITE (*,FMT=*)
      WRITE (*,FMT=99998)
     *  'Original sequence as restored by inverse transform:'
      WRITE (*,FMT=*)
      WRITE (*,FMT=99998) '            Original            Restored'
      IX = 1
      DO 240 I = 1, N
         WRITE (*,FMT=*)
         WRITE (*,FMT=99997) 'Third dimension index = ', I
         DO 220 J = 1, M
            WRITE (*,FMT=99997) 'Column Number ', J
            WRITE (*,FMT=99998)
     *        '          Real    Imag        Real    Imag'
            DO 200 K = 1, L
               WRITE (*,FMT=99999) K, XX(IX), X(IX)
               IX = IX + 1
  200       CONTINUE
  220    CONTINUE
  240 CONTINUE
*
99999 FORMAT (1X,I3,2(:3X,'(',F8.4,',',F8.4,')'))
99998 FORMAT (A)
99997 FORMAT (A,I3)
      END
