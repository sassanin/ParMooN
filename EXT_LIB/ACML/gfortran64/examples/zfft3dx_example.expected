ACML example: FFT of a complex sequence using ZFFT3DX
-----------------------------------------------------

Components of discrete Fourier transform:


Third dimension index =   1
 Column Number   1
          Real    Imag
   1   ( 20.5396,  0.9982)
   2   (  0.5627, -4.2139)
   3   (  0.5466, -0.5738)
   4   (  0.5445, -0.0999)
   5   (  0.5429,  0.2602)
 Column Number   2
          Real    Imag
   1   ( -0.7457,  0.4031)
   2   (  3.1479,  0.4203)
   3   (  0.4287,  0.4083)
   4   (  0.0746,  0.4067)
   5   ( -0.1944,  0.4055)

Third dimension index =   2
 Column Number   1
          Real    Imag
   1   ( -3.3745,  1.8242)
   2   ( 14.2451,  1.9021)
   3   (  1.9399,  1.8477)
   4   (  0.3376,  1.8406)
   5   ( -0.8795,  1.8352)
 Column Number   2
          Real    Imag
   1   ( -1.3627, -2.5208)
   2   ( -1.4209, 10.6414)
   3   ( -1.3803,  1.4491)
   4   ( -1.3750,  0.2522)
   5   ( -1.3709, -0.6570)

Original sequence as restored by inverse transform:

            Original            Restored

Third dimension index =   1
Column Number   1
          Real    Imag        Real    Imag
   1   (  1.5403,  0.8415)   (  1.5403,  0.8415)
   2   (  0.5839,  0.9093)   (  0.5839,  0.9093)
   3   (  0.0100,  0.1411)   (  0.0100,  0.1411)
   4   (  0.3464, -0.7568)   (  0.3464, -0.7568)
   5   (  1.2837, -0.9589)   (  1.2837, -0.9589)
Column Number   2
          Real    Imag        Real    Imag
   1   (  1.9602, -0.2794)   (  1.9602, -0.2794)
   2   (  1.7539,  0.6570)   (  1.7539,  0.6570)
   3   (  0.8545,  0.9894)   (  0.8545,  0.9894)
   4   (  0.0889,  0.4121)   (  0.0889,  0.4121)
   5   (  0.1609, -0.5440)   (  0.1609, -0.5440)

Third dimension index =   2
Column Number   1
          Real    Imag        Real    Imag
   1   (  1.0044, -1.0000)   (  1.0044, -1.0000)
   2   (  1.8439, -0.5366)   (  1.8439, -0.5366)
   3   (  1.9074,  0.4202)   (  1.9074,  0.4202)
   4   (  1.1367,  0.9906)   (  1.1367,  0.9906)
   5   (  0.2403,  0.6503)   (  0.2403,  0.6503)
Column Number   2
          Real    Imag        Real    Imag
   1   (  0.0423, -0.2879)   (  0.0423, -0.2879)
   2   (  0.7248, -0.9614)   (  0.7248, -0.9614)
   3   (  1.6603, -0.7510)   (  1.6603, -0.7510)
   4   (  1.9887,  0.1499)   (  1.9887,  0.1499)
   5   (  1.4081,  0.9129)   (  1.4081,  0.9129)
